lib/stack_master/aws_driver/cloud_formation.rb

```diff
@@ -3,9 +3,15 @@ module AwsDriver
     class CloudFormation
       extend Forwardable
 
-      def set_region(region)
-        @region = region
-        @cf = nil
+      def region
+        @region ||= ENV['AWS_REGION'] || Aws.config[:region] || Aws.shared_config.region
+      end
+
+      def set_region(value)
+        if region != value
+          @region = value
+          @cf = nil
+        end
```

Comment:
Yeah, this deserves some explanation (:warning: it's a bit long).

This solves a bunch of issues we face when region is not explicitly set. In this situation, given there's no way of getting it from the underlying `Aws::CloudFormation::Client`, we obtain it following exactly the same logic used internally by that class, so the reported region should be the same as the one used internally. 

However, in the unlikely event that AWS decides to update the logic it uses in `Client` in the future, the region used by the `Client` might differ from the region reported by us in `CloudFormation`. To avoid this, we use `CloudFormation#region` also to instantiate the internal `Client` (see below).

`set_region` is a different beast. We obliterate the underlying `Client` when someone calls this method, which is not ideal when we are using the test driver: if we call `set_region` after stubbing stacks and events, all those stubs will be lost and tests will fail.

I could have changed the tests, but my understanding is that calling `set_region` should only change it when it is a different one, thus avoiding unnecessary instantiations.